_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,350 @@
+kafka.network:type=RequestMetrics,name=TotalTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=LocalTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=RemoteTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ThrottleTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ResponseQueueTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ResponseSendTimeMs,request=([\w]+)
+# attributes: Count, 99thPercentile
+```
+
+Likewise the following producer metrics can be helpful, perhaps after broker logs reveal some misbehavior.
+
+```sh
+kafka.producer:type=producer-metrics,client-id=([-.\w]+)
+# attributes: txn-init-time-ns-total, txn-begin-time-ns-total, txn-send-offsets-time-ns-total, txn-commit-time-ns-total, txn-abort-time-ns-total
+```
+
+### Basic example
+
+Let's run a basic example to see how transactions work at the partition level.
+You just need an environment with few command line tools (curl, tar, java, mvn).
+
+This application consumes text messages from the `input-topic`, reverts their content, and sends the result to the `output-topic`.
+The following diagram shows the relevant RPCs.
+
+![txn app](/assets/images/posts/2023-04-14-kafka-txn-app.png)
+
+In the following snippet we start the Kafka cluster, run the application, and create a single transaction.
+For the sake of simplicity, we use a single-node cluster running on localhost, but the example also works with a multi-node cluster.
+
+```sh
+# download the Apache Kafka binary and start a single-node Kafka cluster in background
+$ KAFKA_VERSION="3.4.0" KAFKA_HOME="$(mktemp -d -t kafka.XXXXXXX)" && export KAFKA_HOME
+
+$ curl -sLk "https://archive.apache.org/dist/kafka/$KAFKA_VERSION/kafka_2.13-$KAFKA_VERSION.tgz" \
+  | tar xz -C "$KAFKA_HOME" --strip-components 1
+
+$ $KAFKA_HOME/bin/zookeeper-server-start.sh -daemon $KAFKA_HOME/config/zookeeper.properties \
+  && sleep 5 && $KAFKA_HOME/bin/kafka-server-start.sh -daemon $KAFKA_HOME/config/server.properties
+
+# open a new terminal, get and run the application (Ctrl+C to stop)
+$ git clone git@github.com:fvaleri/examples.git -n --depth=1 --filter=tree:0 && cd examples \
+  && git sparse-checkout set --no-cone kafka/kafka-txn && git checkout 957981f5d8ede82a265f695d9a10c83a50af718f
+...
+HEAD is now at 957981f Only reset affected partitions
+
+$ export BOOTSTRAP_SERVERS="localhost:9092" INSTANCE_ID="kafka-txn-0" \
+  GROUP_ID="my-group" INPUT_TOPIC="input-topic" OUTPUT_TOPIC="output-topic"
+
+$ mvn clean compile exec:java -q -f kafka/kafka-txn/pom.xml
+Starting instance kafka-txn-0
+Created topics: input-topic
+Waiting for new data
+...
+
+# come back to the previous terminal and send a test message
+$ $KAFKA_HOME/bin/kafka-console-producer.sh --bootstrap-server :9092 --topic input-topic
+>test
+>^C
+```
+
+All of this is pretty normal, but now comes the interesting part.
+
+How can we identify which partitions are involved in this transaction and inspect their content?
```

Comment:
partitions of the consumer offsets and transaction state topics, right? Isn't it better to clarify? The goal is not pretty clear to me ...

Reply:
No, the question is about every involved partition. In addition to the ones you mention, there is also the output topic partitions.